### PR TITLE
This would give each worker a unique id and matching scope

### PR DESCRIPTION
**Pull request by jhford**
This is instead of using instance id or spot request id.

### CODE REVIEW

File: src/api-v1.js [hunk 1]
```diff
@@ -658,13 +659,17 @@ api.declare({
       token: token,
     });
 
+    let workerId = slugid.v4();
+    let workerGroup = 'ec2_' + secret.region;
```
Comment (djmitche) [inline src/api-v1.js]: I think the workerGroup should be `aws-provisioner` (maybe with `-v1`).  Without the instance ID, the region doesn't make a lot of sense.

**Comment by djmitche** [comment]
This seems to be mixed in with a bunch of other refactors that should be broken out into their own commits, if not a different PR.  But the bits about slugIds work for me.

It's worth noting that releng is currently discussing "auditing" the instances on which tasks ran before making a release.  I'm not sure what that would mean or entail, and I bet it's just as easy with provisioner-specific instance identifiers as with AWS identifiers.

**Comment by jonasfj** [comment]
note, that workerIds only have to be unique within the given workerGroup.

I'm curious what is the motivation...

I think it makes sense to use:
```
workerId = "<instanceId>";
workerGroup = "ec-<region>-<availability-zone>";
```
Then aws-provisioner-v1 just gets the scope: "queue:worker-id:ec2-*".


I think think it really matters... And we probably shouldn't build too many things that relies on this pattern. But when looking at a task it's convenient to easily have the instanceId and region.

So I'm curious what is the motivation of changing this? EC2 does reuse instanceIds, but usually not at the same time :)
Note: I don't really opppose this... Just curious why?

**Comment by jhford** [comment]
@djmitche yes, there are a couple of PRs that I wrote this on top and forgot I had done so until after I had already pushed the PR :/  Sorry!

@jonasfj yes, it's handy, but I'd say that the place for that information is somewhere other than the scopes, as long as there are real reasons to do not put it in the scope.  Did you see the email that I wrote to accompany this PR?  EC2 make no guarantees about reuse of the id that I know other than that they can be reused.

We're also not strictly changing from instance ids to uuids, we're changing from a star scope for all worker groups and all instances to a specific per-instance value.

**Comment by jhford** [comment]
I don't know how the worker-id scope is used on the docker worker.  @gregarndt maybe you can speak to this?

For deployment, I was thinking we could do a staged deployment:

1. set the workerId & workerGroup value in the return of the secret value, but still return the creds with star scope
2. update docker-worker to use the value from the secret to perform its operations against the queue
3. set the scope in the credential to be the resolved group and id value

**Comment by gregarndt** [comment]
Doing a quick search within docker-worker it seems that it's set once during config loading, right now either specified through config or pulling from aws metadata, and then that value is used throughout the worker so I don't see any issue once we have the worker ID from somewhere.

As long as we record somewhere the instance ID from AWS for debugging purposes, I think things will be fine.

**Comment by djmitche** [comment]
The reason not to use instanceId as workerId is that we are not guaranteed to ever get an instanceId from the AWS API (by a strict reading of "eventually consistent", and @jhford  and @rail have both seen the API not have an instanceId even after the instance has started).

I have frequently copy-pasted the instanceId out of a task inspector and searched for that in papertrail in order to find logs from that instance.  Switching to random identifiers would prevent that.

I talked to @escapewindow about auditing release graphs.  For forensic and auditing purposes, releng would like to be able to retrospectively determine "where" things ran.  For auditing purposes, a random ID is probably fine, as it could be correlated with other tasks run on on that same workerId.  For forensics, we would want an instance Id to reference AWS and papertrail logs, but that would be afterward.  So, yes, as long as there's some record eventually clearly linking a random identifier to an instanceId, I think this change would be OK.

**Comment by jonasfj** [comment]
> The reason not to use instanceId as workerId is that we are not guaranteed to ever get an instanceId from the AWS API

I buy that argument, and is at the same time very sad...

Note: workerGroup/workerId is used when claiming tasks. It ensures that runs are somehow linked to the worker they ran on. The worker only cares in the sense that it must send them to the queue..
That said it might make sense for the worker to tag sentry submissions with workerGroup/workerId, but that's a different story.

**Comment by jhford** [comment]
Closed in favour of an patch which we'd consider merging (https://github.com/taskcluster/aws-provisioner/pull/81) for bug 1272643

**Pull request closed without merging**